tough/src/transport.rs

```diff
@@ -1,36 +1,215 @@
-use std::io::Read;
+#[cfg(feature = "http")]
+use crate::ClientSettings;
+#[cfg(feature = "http")]
+use crate::HttpTransport;
+use serde::export::Formatter;
+use std::fmt::{Debug, Display};
+use std::io::{ErrorKind, Read};
 use url::Url;
 
 /// A trait to abstract over the method/protocol by which files are obtained.
-pub trait Transport {
-    /// The type of `Read` object that the `fetch` function will return.
-    type Stream: Read;
+///
+/// The trait hides the underlying types involved by returning the `Read` object as a
+/// `Box<dyn Read>` and by requiring concrete type [`TransportError`] as the error type.
+///
+pub trait Transport: Debug {
+    /// Opens a `Read` object for the file specified by `url`.
+    fn fetch(&self, url: Url) -> Result<Box<dyn Read + Send>, TransportError>;
 
-    /// The type of error that the `fetch` function will return.
-    type Error: std::error::Error + Send + Sync + 'static;
+    /// Returns a clone of `self` as a `Box<dyn Transport>`.
+    ///
+    /// # Why
+    ///
+    /// The [`Repository`] object holds a `Box<dyn Transport>`, and because we want the `Repository`
+    /// object to implement `Clone`, we need a way of cloning the boxed `Transport` without knowing
+    /// its underlying type. We cannot require require the `Clone` trait bound on `Transport`
+    /// because, if we do, the trait can no longer serve as a 'trait object'.
```

Comment:
```suggestion
    /// its underlying type. We cannot require the `Clone` trait bound on `Transport`
    /// because, if we do, the trait can no longer serve as a 'trait object'.
```